core/src/main/java/zipkin2/reporter/HttpEndpointSupplier.java

```diff
@@ -0,0 +1,115 @@
+ * <ul>
+ *   <li>During build, the sender should invoke the {@linkplain Factory}.</li>
+ *   <li>If the result is {@link Constant}, build the sender to use a static value.</li>
+ *   <li>Otherwise, call {@link HttpEndpointSupplier#get()} each time
+ *       {@linkplain BytesMessageSender#send(List)} is invoked.</li>
+ *   <li>Call {@link #close()} once during {@link BytesMessageSender#close()}.</li>
+ * </ul>
+ *
+ * <h3>Implementation Notes</h3>
+ *
+ * <p>Implement friendly {@code toString()} functions, that include the real endpoint or the one
+ * passed to the {@linkplain Factory}.
+ *
+ * <p>Senders are not called during production requests, rather in time or size bounded loop, in a
+ * separate async reporting thread. Implementations that resolve endpoints via remote calls, such as
+ * from Eureka, should cache internally to avoid blocking the reporter thread on each loop.
+ *
+ * <p>Some senders, such as Armeria, may have more efficient and precise endpoint group logic. In
+ * scenarios where the sender is known, interfaces here may be used as markers. Doing so can satisfy
+ * dependency injection, without limiting an HTTP framework that can handle groups, to a
+ * single-endpoint supplier.
+ *
+ * @since 3.3
+ */
+public interface HttpEndpointSupplier extends Closeable {
+  /**
+   * HTTP {@link BytesMessageSender sender} builders check for this symbol, and return the input as
+   * a {@linkplain Constant} result rather than perform dynamic lookups.
+   *
+   * @since 3.3
+   */
+  Factory CONSTANT_FACTORY = new Factory() {
+    @Override public Constant create(String endpoint) {
+      return new Constant(endpoint);
+    }
+  };
+
+  /**
+   * Returns a possibly cached endpoint to an HTTP {@link BytesMessageSender sender}.
+   *
+   * <p>This will be called inside {@linkplain BytesMessageSender#send(List)}, unless this is an
+   * instance of {@linkplain Constant}.
+   *
+   * @since 3.3
+   */
+  String get();
+
+  /**
+   * Factory passed to HTTP {@link BytesMessageSender sender} builders to control resolution of the
+   * static endpoint from configuration.
+   *
+   * <p>Invoke this when building a sender, not during {@linkplain BytesMessageSender#send(List)}.
+   *
+   * @since 3.3
+   */
+  interface Factory {
+
+    /**
+     * Returns a possibly {@linkplain Constant} endpoint supplier, given a static endpoint from
+     * configuration.
+     *
+     * <p>Note: Some factories may perform I/O to lazy-create a {@linkplain Constant} endpoint.
+     *
+     * @param endpoint a static HTTP endpoint from configuration. For example,
+     *                 http://localhost:9411/api/v2/spans
+     */
+    HttpEndpointSupplier create(String endpoint);
+  }
+
+  /**
+   * HTTP {@link BytesMessageSender senders} check for this type, and will cache its first value.
+   *
+   * @since 3.3
+   */
+  final class Constant implements HttpEndpointSupplier {
+    private final String endpoint;
+
+    public Constant(String endpoint) {
```

Comment:
Related to above, this looks like the "real" public API, a static method may encapsulate better

Reply:
ok I'll move this out to a different file and make the function. no problem.

Reply:
actually I can just hide the ctor, we still need the type public

Reply:
actually if I move things I have to move everything to a new type, and now the interface has to refer to things in another file to talk about behavior. this is unlike the rest of the codebase, but I'll raise a PR and you can decide.